src/tasks/python/PythonExtensionHelper.ts

```diff
@@ -0,0 +1,116 @@
+
+export namespace PythonExtensionHelper {
+    export interface DebugLaunchOptions {
+        host?: string;
+        port?: number;
+        wait?: boolean;
+    }
+
+    export function getDebuggerEnvironmentVars(): { [key: string]: string } {
+        return { ["PTVSD_LOG_DIR"]: `/dbglogs` };
+    }
+
+    export function getDebuggerLogFilePath(folderName: string) : string {
+    // The debugger generates the log file with the name in this format: ptvsd-{pid}.log,
+    // So given that we run the debugger as the entry point, then the PID is guaranteed to be 1.
+        return path.join(os.tmpdir(), folderName, "ptvsd-1.log");
+    }
+
+    export async function ensureDebuggerReady(prelaunchTask: vscode.Task, debuggerSemaphorePath: string, containerName: string, cliDockerClient: CliDockerClient): Promise<void> {
+        // tslint:disable-next-line:promise-must-complete
+        return new Promise((resolve, reject) => {
+            vscode.tasks.onDidEndTask(async e => {
+                if (e.execution.task === prelaunchTask) {
+                    // There is no way to know the result of the completed task, so a best guess is to check if the container is running.
+                    const containerRunning = await cliDockerClient.inspectObject(containerName, { format: "{{.State.Running}}" });
+
+                    if (containerRunning === "false") {
+                        reject("Failed to attach the debugger, please see the terminal output for more details.");
+                    }
+
+                    const maxRetriesCount = 20;
+                    let retries = 0;
+                    let created = false;
+
+                    // Look for the magic string below in the log file with a retry every 0.5 second for a maximum of 10 seconds.
+                    // TODO: Should be gone as soon as the retry logic is part of the Python debugger/extension.
+                    while (++retries < maxRetriesCount && !created) {
+                        if (fse.existsSync(debuggerSemaphorePath)) {
+                            const contents = fse.readFileSync(debuggerSemaphorePath);
+
+                            created = contents.toString().indexOf("Starting server daemon on") >= 0;
+                            if (created) {
+                                break;
+                            }
+                        }
+
+                        await delay(500);
+                    }
+
+                    if (created) {
+                        resolve();
+                    } else {
+                        reject("Failed to attach the debugger within the alotted timeout.");
+                    }
+                }});
+        })
+    }
+
+    export function getRemotePtvsdCommand(target: PythonFileTarget | PythonModuleTarget, args?: string[], options?: DebugLaunchOptions): string {
+        let fullTarget: string;
+
+        if ((target as PythonFileTarget).file) {
```

Comment:
Are the casts needed?  Usually TS is smart enough to notice that if a property exclusive to a type exists, it then "becomes" that type.  Sometimes a ["type guard"](https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-the-in-operator) is used to help TS identify a type.  You may be able to do something like:

```typescript
if ('file' in target) {
  // Target is now of type PythonFileTarget
} else if ('module' in target) {
  // Target is now of type PythonModuleTarget
} else {
  // Target is of unknown type
}
```